app/scripts/lib/crypto/recovery-keys.js

```diff
@@ -0,0 +1,175 @@
+
+define(function (require, exports, module) {
+  const sjcl = require('sjcl');
+
+  function importFxaCryptoDeriver() {
+    return import(/* webpackChunkName: "fxaCryptoDeriver" */ 'fxaCryptoDeriver');
+  }
+
+  const RECOVERY_KEY_VERSION = '\x01';
+  const ALPHABET_BASE32 = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';
+
+  return {
+    /**
+     * Generate a random base32 string of length and
+     * prepend encryption version number.
+     *
+     * @param {Integer} length Length of string to generate
+     * @returns {string}
+     */
+    generateRecoveryKey: function (length) {
+      return importFxaCryptoDeriver().then((fxaCryptoDeriver) => {
+        const jose = fxaCryptoDeriver.jose;
+        const bytes = jose.util.randomBytes(length - 1);
+        const code = [];
+        for (let i = 0; i < length; i++) {
+          code.push(ALPHABET_BASE32[bytes[i] % 32]);
+        }
+
+        // Prepend version and return code
+        return RECOVERY_KEY_VERSION + code.join('');
+      });
+    },
+
+    /**
+     * Creates the encryption keys from the uid and recovery key. The object contains
+     * `recoveryKeyId` which can be used to retrieve the recovery data and
+     * `recoveryEnc` which can be used to decrypt the data.
+     *
+     * @param {String} uid Uid of user
+     * @param {String} recoveryKey Base32 encoded string
+     * @returns {Promise} A promise that will be fulfilled with the `recoveryKeyId` and `recoveryEnc`
+     */
+    getRecoveryEncryption: function (uid, recoveryKey) {
+      return Promise.resolve()
+        .then(() => {
+          if (! uid) {
+            throw new Error('Missing uid');
+          }
+
+          if (! recoveryKey) {
+            throw new Error('Missing recoveryKey');
+          }
+
+          const km = sjcl.codec.base32.toBits(recoveryKey);
+          const salt = sjcl.codec.hex.toBits(uid);
+
+          const info1 = sjcl.codec.utf8String.toBits('fxa recovery fingerprint');
+          const info2 = sjcl.codec.utf8String.toBits('fxa recovery encrypt key');
+
+          const recoveryKeyIdRaw = sjcl.misc.hkdf(km, 16 * 8, salt, info1);
+          const recoveryKeyEncRaw = sjcl.misc.hkdf(km, 32 * 8, salt, info2);
+
+          // Convert to hex strings for readability
+          const recoveryKeyId = sjcl.codec.hex.fromBits(recoveryKeyIdRaw);
+          const recoveryEnc = sjcl.codec.hex.fromBits(recoveryKeyEncRaw);
+
+          return {
+            recoveryEnc,
+            recoveryKeyId
+          };
+        });
+    },
+
+    /**
+     * Creates an encrypted bundle for the recovery data provided using
+     * the recovery key id and recovery encryption.
+     *
+     * This encryption is based on the protocol described in
+     *
+     * https://github.com/mozilla/fxa-auth-server/blob/master/docs/recovery_keys.md
+     *
+     * @param {String} recoveryKeyId Recovery key id used in encryption
+     * @param {String} recoveryEnc Recovery encryption used to encrypt bundle
```

Comment:
Not being part of the crypto world, I don't know what `Recovery encryption` means. The method below this describes this as "recovery encryption key" If it's a key, can you rename the variable `recoveryEncKey` for clarity and update the description?

Reply:
I just saw that @rfk suggested returning a JWK from `getRecoveryEncryption` so this point would be moot, `recoveryEn` would be `recoveryJWK` or similar.